### PR TITLE
"Required Changes" Branch

### PR DESCRIPTION
**Pull request by 300ms**
This is feature branch to get a code review for if all requested changes have been made or not.

### CODE REVIEW

File: enumerable_methods.rb [hunk 1]
```diff
@@ -55,14 +59,16 @@ def my_select
...
       x = true
       i = 0
       while i < arr.length
-        x = false unless yield(arr[i])
+        if block_given?
+          x = false unless yield(arr[i])
+        else
+          x = false unless arr[i]
+        end
         i += 1
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* When a class is passed as an argument returns true if all of the collection is a member of such class.
```
Failure/Error: expect(array.my_all?(Integer)).to be array.all?(Integer)
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
```
* When a Regex is passed as an argument returns true if all of the collection matches the Regex.
```
Failure/Error: expect(words.my_all?(/d/)).to be words.all?(/d/)
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
```
* When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern.
```
Failure/Error: expect(array.my_all?(3)).to be array.all?(3)
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
```

File: enumerable_methods.rb [hunk 2]
```diff
@@ -72,14 +78,15 @@ def my_all?
...
-        return true if yield(arr[i])
-
+        if block_given?
+          return true if yield(arr[i])
+        else
+          return true if arr[i]
+        end
         i += 1
       end
       false
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* When a class is passed as an argument returns true if at least one of the collection is a member of such class.
```
Failure/Error: expect(array.my_any?(Integer)).to be array.any?(Integer)
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
```
* When a Regex is passed as an argument returns false if none of the collection matches the Regex.
```
Failure/Error: expect(words.my_any?(/z/)).to be words.any?(/z/)
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
```
* When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern.
```
Failure/Error: expect(words.my_any?('cat')).to be words.any?('cat')
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
```

File: enumerable_methods.rb [hunk 3]
```diff
@@ -89,14 +96,16 @@ def my_any?
...
       while i < arr.length
-        x = false if yield(arr[i])
+        if block_given?
+          x = false if yield(arr[i])
+        else
+          x = false if arr[i]
+        end
         i += 1
       end
       x
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* When a class is passed as an argument returns true if none of the collection is a member of such class.
```
Failure/Error: expect(array.my_none?(String)).to be true
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
```
* When a Regex is passed as an argument returns true only if none of the collection matches the Regex.
```
Failure/Error: expect(words.my_none?(/z/)).to be words.none?(/z/)
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
```
* When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern.
```
Failure/Error: expect(words.my_none?(5)).to be words.none?(5)
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
```

File: enumerable_methods.rb [hunk 4]
```diff
@@ -106,14 +115,16 @@ def my_none?
...
       while i < arr.length
-        x += 1 if yield(arr[i])
+        if block_given?
+          x += 1 if yield(arr[i])
+        else
+          x += 1 if arr[i]
+        end
         i += 1
       end
       x
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* Counts the number of items in enum that are equal to item if an argument is given.
```
Failure/Error: expect(array.my_count(LOWEST_VALUE)).to eq array.count(LOWEST_VALUE)
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
```

File: enumerable_methods.rb [hunk 5]
```diff
@@ -173,3 +184,7 @@ def my_map(prc)
 end
 
 # rubocop: enable Metrics/ModuleLength
```
Comment (wathigo) [inline enumerable_methods.rb:196]: ## my_map
**[REQUIRED]**

* Returns a new array with the results of running block once for every element in enum.
```
Failure/Error: expect(array.my_map(&block)).to eq array.map(&block)
      
      ArgumentError:
        wrong number of arguments (given 0, expected 1)
```
* Returns an Enumerator if no block is given
```
Failure/Error: expect(array.my_map).to be_an(Enumerator)
      
      ArgumentError:
        wrong number of arguments (given 0, expected 1)
```

File: enumerable_methods.rb [hunk 6]
```diff
@@ -123,14 +134,16 @@ def my_count
...
       i = 1
       while i < arr.length
-        x = yield(x, arr[i])
+        x = if block_given?
+              yield(x, arr[i])
+            else
+              [x, arr[i]]
+            end
         i += 1
       end
```
Comment (wathigo) [inline enumerable_methods.rb]: **[REQUIRED]**
* Combines all elements of enum by applying a binary operation, specified by a block.
```
Failure/Error: actual = range.my_inject(4) { |prod, n| prod * n }
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
```
* When a symbol is specified combines each element of the collection by applying the symbol as a named method.
```
Failure/Error: expect(array.my_inject(:+)).to eq array.inject(:+)
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
```

**Review by wathigo — CHANGES_REQUESTED** [review]
## Changes Required 

Hi Devs :wave: ,
Great work implementing the [Git Flow](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/).
Let's refer to [Ruby docs](https://ruby-doc.org/core-2.6.5/Enumerable.html) for the correct behaviour of the methods.
Let's fix the issues pointed out in the comments :point_down:  

### After implementing the changes, submit another request for review.

Feel free to reach out via slack for any issue/clarification.

_Happy Coding :computer: 
Reviewed by: **Simon Wathigo**_

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good job guys :+1: 
You implemented some of the suggested changes. That's good. However, we can not merge this PR yet because some edge cases are still failing. Look at my comments and do your best to implement them then resubmit for another review.

### Martin Afani

File: enumerable_methods.rb [hunk 7]
```diff
@@ -54,94 +66,147 @@ def my_select
...
+    else
       x = true
       i = 0
+      cls.class == Integer.class ? cls = cls : cls = cls.class
+      puts cls
       while i < arr.length
-        x = false unless yield(arr[i])
+        x = false unless arr[i].class == cls
         i += 1
       end
```
Comment (whiz25) [inline enumerable_methods.rb]: - [x] when a class is passed as an argument returns true if all of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerable_methods.rb [hunk 8]
```diff
@@ -54,94 +66,147 @@ def my_select
...
-        return true if yield(arr[i])
+        return true if arr[i].class == cls
 
         i += 1
       end
       false
-    else
-      puts 'Error: Parameter is not an array!'
     end
   end
```
Comment (whiz25) [inline enumerable_methods.rb]: - [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerable_methods.rb [hunk 9]
```diff
@@ -54,94 +66,147 @@ def my_select
...
       while i < arr.length
-        x = false if yield(arr[i])
+        x = false if arr[i].class == cls
         i += 1
       end
       x
-    else
-      puts 'Error: Parameter is not an array!'
     end
   end
```
Comment (whiz25) [inline enumerable_methods.rb]: - [x] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable_methods.rb [hunk 10]
```diff
@@ -152,24 +217,33 @@ def multiply_els
...
+        x.push(prc.call(arr[i]))
       end
-      x
-    else
-      puts 'Error: Parameter is not an array!'
+      i += 1
     end
+    x
   end
 end
```
Comment (whiz25) [inline enumerable_methods.rb]: - [x] returns an Enumerator if no block is given

File: enumerable_methods.rb [hunk 11]
```diff
@@ -54,94 +66,147 @@ def my_select
...
+    while i < arr.length
+      x = if block_given?
+            yield(x, arr[i])
+          else
+            [x, arr[i]]
+          end
+      i += 1
     end
+    x
   end
```
Comment (whiz25) [inline enumerable_methods.rb]: - [x] combines all elements of enum by applying a binary operation, specified by a block
- [x] when a symbol is specified combines each element of the collection by applying the symbol as a named method

**Review by whiz25 — APPROVED** [review]
## Nice try
I know have tried your best to implement the changes but I must say all changes haven't been implemented :disappointed:. This review has reached the max number of cycles so it's a :+1: from me.

### Martin Afani

**Pull request merged**